tiledb/sm/storage_manager/consolidator.cc

```diff
@@ -590,18 +513,31 @@ Status Consolidator::create_queries(
     Array* array_for_reads,
     Array* array_for_writes,
     bool sparse_mode,
-    void* subarray,
+    const NDRange& subarray,
     void** buffers,
     uint64_t* buffer_sizes,
     Query** query_r,
     Query** query_w,
     URI* new_fragment_uri) {
+  // Create subarray to set
+  std::vector<uint8_t> s;
+  if (!sparse_mode) {
+    auto domain_size = 2 * array_for_reads->array_schema()->coords_size();
+    s.resize(domain_size);
+    uint64_t offset = 0;
+    for (const auto& r : subarray) {
+      std::memcpy(&s[offset], r.data(), r.size());
+      offset += r.size();
+    }
+  }
+  void* to_set = (s.empty()) ? nullptr : &s[0];
+
   // Create read query
   *query_r = new Query(storage_manager_, array_for_reads);
   RETURN_NOT_OK((*query_r)->set_layout(Layout::GLOBAL_ORDER));
   RETURN_NOT_OK(
       set_query_buffers(*query_r, sparse_mode, buffers, buffer_sizes));
-  RETURN_NOT_OK((*query_r)->set_subarray(subarray, true));
+  RETURN_NOT_OK((*query_r)->set_subarray(to_set, true));
```

Comment:
This doesn't look safe me because `to_set` will be freed when this routine returns (because `vector<uint8_t> s` will destruct).

I traced `set_subarray` to see if it would copy `to_set`. It appears that it does not copy and will store pointers into `to_set` on this line in `Subarray::add_range`: `ranges_[dim_idx].add_range(range);`

Reply:
`ranges_[dim_idx].add_range(range)` (i.e., `Subarray::Ranges::add_range`) copies `range` to its internal buffer.

Reply:
Thanks! I should have dug a little deeper.